### PR TITLE
Correction in tone mapping shader for transparent object

### PR DESCRIPTION
**Pull request by alja**
This PR fixed fixes a problem for transparent object with a white background.
It also includes one commit for RenderCore distribution in ROOT project.

### CODE REVIEW

**Comment by alja** [comment]
Example of 98% transparent box before and after the correction. 
Note: the change also needed to be done in the render passes. The backforund/clear color was set black in all the render passes except the tone mapper render pass.

Box with transparency of 98% before the correction
Screenshot from 2024-05-13 16-00-25


Box with transparency of 98% after the correction in tone mapping
Screenshot from 2024-05-13 15-53-05

**Comment by ss4774** [comment]
Were these example pictures made after the code was already changed? Would it be please possible to get another picture of this box (after the tone mapping correction), but before the changes to the shader code were made?

**Comment by osschar** [comment]
Hi, the problem was ... we used white clear-color for prior passes ... so even if the alpha channel was very low, the white would "shine" through and get converted to a gray.
Now we have transitioned to using (0,0,0,0) as clear color everywhere else ... and the background color is only applied in the final tone-mapper.
I think there is no other way to this properly.

**Comment by alja** [comment]
@ss4774 The top photo is the 98% transparent box before the changes. It looks grey instead of almost completely sheer.
The bottom photo is a screenshot of 98% transparent box after the correction.
Before the correction even 100% transparent box looked gray.

**Comment by alja** [comment]
# Screenshot of 95% transparent geometry using RenderCore engine

Screen Shot 2024-05-07 at 11 30 23 AM
# Screenshot of 95% transparent geometry using Tree.js engine
Screen Shot 2024-05-07 at 11 34 38 AM

**Comment by ss4774** [comment]
As you have pointed out, the box in the first picture should be less visible if the opacity is set to $\alpha = (1 - 0.98) = 0.02$. Also, the culling appears to be set for back faces.

**Comment by ss4774** [comment]
A possible solution might be to use a clear color with really high values (e.g. $100$ or maybe even $1000$ per RGB channel). The tone mapping should map these high values close to 1.0.
Another way might also be to apply tone mapping normally and then shift/stretch the entire buffer so that the dim/light background value reaches the range of $[0.0 - 1.0]$. However, this would require you to find the lowest/highest tone-mapped value, which is not easily done without the compute shaders.

**Comment by ss4774** [comment]
The original tone mapping shader code uses the $\alpha$ channel as a hard switch, to apply pure clear color whenever the $\alpha$ is strictly equal to $0$. Your solution looks much more appropriate. I would however remove the whole `if else` statement (I will set a comment in the source code). I am happy with the code as it is, but would try to avoid branches. Let me know if you want me to do this.

File: src/shaders/custom/post_process/ToneMapping.frag [hunk 1]
```diff
@@ -49,8 +49,8 @@ void main() {
         // gamma correction 
         mapped = pow(mapped, vec3(1.0 / gamma));
     
-        if (hdrColor.a == 0.0)
-            ldrColor = vec4(u_clearColor.rgb, 1.0);
+        if (hdrColor.a < 1.0)
+            ldrColor = vec4(mapped * hdrColor.a + u_clearColor.rgb * (1.0 - hdrColor.a), 1.0);
         else
```
Comment (ss4774) [inline src/shaders/custom/post_process/ToneMapping.frag]: I would replace the whole `if else` statement and only keep the single `ldrColor` evaluation (and maybe clamp the `hdrColor.a` value).
Or even better, use the `mix` function:
```glsl
ldrColor = vec4(mix(u_clearColor.rgb, mapped, clamp(hdrColor.a, 0.0, 1.0)), 1.0);
```

**Comment by alja** [comment]
@ss4774 I have used the mix function. Thank you for the suggestion!

**Comment by alja** [comment]
> A possible solution might be to use a clear color with really high values (e.g. 100 or maybe even 1000 per RGB channel). The tone mapping should map these high values close to 1.0. Another way might also be to apply tone mapping normally and then shift/stretch the entire buffer so that the dim/light background value reaches the range of [0.0−1.0]. However, this would require you to find the lowest/highest tone-mapped value, which is not easily done without the compute shaders.

We were discussing compute shaders to access the max val as well.
BTW, I also tried extended reinhart [1], where I put the maximum value as a number of light in scene (5 lights in case of REve 3D viewer). The difference between plain and extended was barely visible. I was  testing tone mapping using the exposure function in FireworksWeb. It works better for some scenes. I will add interface (e.g. dataGUI) to set the method and its parameters.   

[1] https://64.github.io/tonemapping/

**Comment by ss4774** [comment]
Glad to be of help! Merging the pull request.

**Pull request merged**